### PR TITLE
Teamcity: Fail if yarn outputs specific warnings

### PR DESCRIPTION
**Pull request by noahtallen**
### Changes proposed in this Pull Request
I've noticed that when I'm reviewing renovate PRs, I check `yarn install` locally to verify that peer dependencies are all good. In some cases, we have to make some updates if one of two warnings occurs: `YN0002` (unmet peer dep) and `YN0068` (related to yarnrc.yml overrides, normally when we update a package such that the override no longer matches the version) 

Why not incorporate those checks into TeamCity?

This will fail the unit test build if either YN0002 or YN0068 occurs during yarn install.

Screen Shot 2021-11-16 at 7 19 08 PM

### Testing instructions
- [x] Incorporate a warning related to peer deps. Unit test should fail
- [x] Incorporate a warning related to yarnrc. Unit test should fail
- [x] Resolve warnings. Unit test should pass
- [x] Tests other than unit tests should pass regardless (pretty sure the failures at the time were just flaky)

### CODE REVIEW

**Comment by scinos** [comment]
Instead of checking this using shell script + TC messages, what do you think about using `logFilters` to elevate severity of YN0002 and YN0068 to `error`?

**Comment by noahtallen** [comment]
Cool, I didn't even think of that! :) It definitely works, testing locally.

My two thoughts about it are:
- It might still be nice to use TC messages to make failing CI easier to troubleshoot
- Do we actually want these errors to fail install locally? For example, YN0068 is just saying you have an entry in yarnrc.yml which doesn't match a dependency you install. While we want to prevent merging PRs which introduce that message, it's really just a warning locally

**Comment by scinos** [comment]
Good point about having a better message to help to troubleshoot. But I'd be wary about adding complexity to solve a problem that we don't know if we have. Maybe most folks that encounter this problem find the default yarn message enough to solve it.

About failing locally... I'm not sure, to be honest. For some folks, it may be annoying if you are just testing some new package or config, and `yarn` forces you to fix the warning right there. But some other users may find it annoying that `yarn` allows you to get with some action that we know it will ultimately fail in CI. My _guess_ is that the first group is smaller and the folks there can work around the limitation easily.

We can go with the simplest solution (elevate the error level and let yarn fail) and monitor our slack channels and see if that causes confusion.

At any rate, if we decide to go implement better error messaging, I'd suggest having a look at `yarn install --json`. The theory is that `--json` makes the output easier and safer to parse by machines (actually, now that I think about it, it would be really cool to have some sort of "generic TeamCity reporter" that can listen to `--json` output and report warnings/errors to TeamCity using the service message format)

**Comment by noahtallen** [comment]
Those are good points! Moving forward with the log-level implementation now, and testing it out in CI.

**Comment by noahtallen** [comment]
I'll leave the PR in this state as an example of what it would look like if it fails. IMO, the previous behavior was a lot easier to troubleshoot. I think it takes a while longer to find the problem with this approach.

**Comment by noahtallen** [comment]
This is ready to go -- TC should fail because there is an existing peer dep warning on trunk

**Comment by noahtallen** [comment]
I think I still like the original approach better from the troubleshooting point of view, but this is easier to maintain :man_shrugging:

**Comment by scinos** [comment]
> I think I still like the original approach better from the troubleshooting point of view, but this is easier to maintain 🤷‍♂️

Yeah I see what you mean. There is nothing that says this line is causing the build to fail

```
➤ YN0002: │ react-native-codegen@npm:0.0.7 doesn't provide @babel/preset-env (pe4ee6), requested by jscodeshift
```

And there are a ton of lines with `STDERR` in them, these that are red herrings and will confuse people

**Comment by noahtallen** [comment]
I reintroduced the previous behavior

File: .teamcity/_self/projects/WebApp.kt [hunk 1]
```diff
@@ -171,6 +171,22 @@ object RunAllUnitTests : BuildType({
...
+				declare -A fail_yarn
+				fail_yarn[YN0002]="There are unmet peer dependencies."
+				fail_yarn[YN0068]="An entry in yarnrc.yml is out of date."
+
+				for yarn_err_code in "${'$'}{!fail_yarn[@]}"
+				do
+					if echo "${'$'}yarn_output" | grep -q "${'$'}yarn_err_code" ; then
+						err_msg="Yarn error ${'$'}yarn_err_code: ${'$'}{fail_yarn[${'$'}yarn_err_code]}"
+						echo "${'$'}err_msg"
+						echo "##teamcity[buildProblem description='${'$'}err_msg' identity='yarn_problem']]"
```
Comment (scinos) [inline .teamcity/_self/projects/WebApp.kt]: Repeated `]`
```suggestion
						echo "##teamcity[buildProblem description='${'$'}err_msg' identity='yarn_problem']"
```

File: .teamcity/_self/projects/WebApp.kt [hunk 2]
```diff
@@ -171,6 +171,22 @@ object RunAllUnitTests : BuildType({
...
+				# action on them before merging the PR.
+				declare -A fail_yarn
+				fail_yarn[YN0002]="There are unmet peer dependencies."
+				fail_yarn[YN0068]="An entry in yarnrc.yml is out of date."
+
+				for yarn_err_code in "${'$'}{!fail_yarn[@]}"
+				do
+					if echo "${'$'}yarn_output" | grep -q "${'$'}yarn_err_code" ; then
+						err_msg="Yarn error ${'$'}yarn_err_code: ${'$'}{fail_yarn[${'$'}yarn_err_code]}"
+						echo "${'$'}err_msg"
```
Comment (scinos) [inline .teamcity/_self/projects/WebApp.kt]: I think we don't need this line, as it seems to generate a duplicated output:
image

File: .teamcity/_self/yarnInstall.kt [hunk 3]
```diff
@@ -7,6 +7,7 @@ val yarn_install_cmd = """
 	# Do not type-check packages on postinstall, we already have a unit test for that
 	export SKIP_TSC=true
 
-	# Install modules.
-	yarn install
+	# Install modules. Outputs to stdout while also saving to a variable for later use.
+	exec 5>&1
+	yarn_output=${'$'}( yarn install 2>&1 | tee /dev/fd/5 )
```
Comment (scinos) [inline .teamcity/_self/yarnInstall.kt]: We could as well save it to a file, right? I think it may be simpler.

File: .teamcity/_self/projects/WebApp.kt [hunk 4]
```diff
@@ -171,6 +171,22 @@ object RunAllUnitTests : BuildType({
...
+				# Fail when we encounter these yarn warnings. We want to take
+				# action on them before merging the PR.
+				declare -A fail_yarn
+				fail_yarn[YN0002]="There are unmet peer dependencies."
+				fail_yarn[YN0068]="An entry in yarnrc.yml is out of date."
+
+				for yarn_err_code in "${'$'}{!fail_yarn[@]}"
+				do
+					if echo "${'$'}yarn_output" | grep -q "${'$'}yarn_err_code" ; then
+						err_msg="Yarn error ${'$'}yarn_err_code: ${'$'}{fail_yarn[${'$'}yarn_err_code]}"
```
Comment (scinos) [inline .teamcity/_self/projects/WebApp.kt]: It would be cool if we can also add the original error line.

Instead of trying to extract it from the original output, we could run `yarn` again with the flag `--json` when we know there was an error, and parse the JSON output

**Review by scinos — APPROVED** [review]

**Comment by noahtallen** [comment]
Thanks for the feedback, I'll address that today!

File: yarn.lock [hunk 5]
```diff
@@ -15763,10 +15817,10 @@ __metadata:
   languageName: node
   linkType: hard
 
-"electron-to-chromium@npm:^1.3.564, electron-to-chromium@npm:^1.3.886":
-  version: 1.3.889
-  resolution: "electron-to-chromium@npm:1.3.889"
-  checksum: 6e50fddd1301f6a79556866c58bee75c51717ee64e562bfeb624453aa52953b5eb0248c118efc27ad3bfe86c75bd57c9cf6f5b2aeb2d7809426cd307093d383b
+"electron-to-chromium@npm:^1.3.564, electron-to-chromium@npm:^1.3.896":
+  version: 1.3.900
```
Comment (scinos) [inline yarn.lock]: Why this version change? Doesn't seem related to the original yarn errors.

File: yarn.lock [hunk 6]
```diff
@@ -11394,18 +11448,18 @@ __metadata:
   languageName: node
   linkType: hard
 
-"browserslist@npm:^4.0.0, browserslist@npm:^4.12.0, browserslist@npm:^4.14.5, browserslist@npm:^4.16.0, browserslist@npm:^4.16.3, browserslist@npm:^4.16.6, browserslist@npm:^4.17.6, browserslist@npm:^4.8.2":
-  version: 4.17.6
-  resolution: "browserslist@npm:4.17.6"
+"browserslist@npm:^4.0.0, browserslist@npm:^4.12.0, browserslist@npm:^4.14.5, browserslist@npm:^4.16.0, browserslist@npm:^4.16.3, browserslist@npm:^4.16.6, browserslist@npm:^4.17.5, browserslist@npm:^4.17.6, browserslist@npm:^4.8.2":
+  version: 4.18.1
```
Comment (scinos) [inline yarn.lock]: nit: We don't really need to bump this one, as the old version will satisfy the new range `browserslist@npm:^4.17.5`.  It may cause more browsers to get the browserhappy redirect.

File: yarn.lock [hunk 7]
```diff
@@ -2782,17 +2800,17 @@ __metadata:
   languageName: node
   linkType: hard
 
-"@babel/preset-env@npm:^7.1.6, @babel/preset-env@npm:^7.12.1, @babel/preset-env@npm:^7.12.11, @babel/preset-env@npm:^7.13.10, @babel/preset-env@npm:^7.16.0":
-  version: 7.16.0
-  resolution: "@babel/preset-env@npm:7.16.0"
+"@babel/preset-env@npm:*, @babel/preset-env@npm:^7.1.6, @babel/preset-env@npm:^7.12.1, @babel/preset-env@npm:^7.12.11, @babel/preset-env@npm:^7.13.10, @babel/preset-env@npm:^7.16.0":
+  version: 7.16.4
```
Comment (scinos) [inline yarn.lock]: nit: We don't really need to update this package, as the existing `7.16.0` will still satisfy all ranges. But I think you'd need to edit the file manually. Up to you.

**Comment by noahtallen** [comment]
The yarn.lock updates were a result of fixing the peer dep warnings on trunk:

1. Adding unfulfilled peer deps to the e2e package (pretty sure these match the existing packages)
2. Adding the resolution for react native codegen
3. Runing `yarn dedupe`

this could be split to a separate PR, but I wanted to test it to make sure this PR works in the happy path too. I'm not sure why those are updating, but that's what happened when I ran `yarn dedupe`

**Comment by scinos** [comment]
> `yarn dedupe`

Ah, that makes sense. I've seen it cause problems in this scenario:

1. Let's say we have a dependency on `lib@^1.0.0` that got resolved as `lib@1.2.3`
2. We install something that has a dependency on `lib@^1.1.0`. As this is a new range, `yarn` adds a new entry in `yarn.lock` and resolves it to the _latest published version_. It doesn't try to use the existing `lib@1.2.3`. Let's say it gets resolved as `lib@1.3.0`.
3. So now we have `lib@^1.0.0 = lib@1.2.3` and `lib@^1.1.0 = lib@1.3.0`. `yarn dedupe` will find the duplicated packages and will complain
4. As `yarn dedupe` only implements the `highest` strategy, will use `lib@1.3.0` for both ranges.
5. As now `lib@1.2.3` is unused, it will get removed from `yarn.lock`.

So effectively, even if the existing `lib@1.2.3` satisfies both existing and new ranges, it will be updated to the latest published version.

**Comment by noahtallen** [comment]
Yep, makes sense! If we want to avoid this behavior, should be we pinning versions instead of using version ranges? If we are saying we support `"browserslist": "^4.17"`, that means we support 4.18 as well, so yarn isn't breaking our contract (so to speak) when it updates it.

**Comment by scinos** [comment]
> Yep, makes sense! If we want to avoid this behavior, should be we pinning versions instead of using version ranges?

I think it will still auto-update third-party dependencies, but I haven't tested it.

You are right that it isn't breaking our contract, and ultimately updating versions is a Good Thing. But I wish it will do it less automatically. If I do `yarn add foo && yarn dedupe`, I don't expect it to end up updating `bar` across the whole repo.
Maybe have a config setting or a flag where I can ask yarn to use my existing dependencies first when possible.

**Comment by noahtallen** [comment]
Yeah, that would be nice.
